### DISCUSSION TITLE
usb.core.USBError: [Errno 32] Pipe error after some time

### DISCUSSION THREAD

**Discussion by danielhrisca** [pyusb/pyusb#415] [Q&A] [answered]
Hello,

I'm getting this error while reading. The communication works fine for about 3 minutes and then this is triggered. 

I'm using an array for the reads, the essential code looks roughly like this:

```python

data_buffer = array("B", bytes(RECV_SIZE))

while True:
    read_count = endpoint.read(data_buffer, 1) 
    if not read_count:
            sleep(0.01)
            continue
    # some data processing here
```

and the traceback looks like this

```
Traceback (most recent call last):
  File "\devices\usbbox.py", line 364, in _data_listen
    read_count = endpoint.read(data_buffer, 1) 
  File "D:\TMP\project_venv_3.9\lib\site-packages\usb\core.py", line 423, in read
    return self.device.read(self, size_or_buffer, timeout)
  File "D:\TMP\project_venv_3.9\lib\site-packages\usb\core.py", line 1029, in read
    ret = fn(
  File "D:\TMP\project_venv_3.9\lib\site-packages\usb\backend\libusb1.py", line 846, in bulk_read
    return self.__read(self.lib.libusb_bulk_transfer,
  File "D:\TMP\project_venv_3.9\lib\site-packages\usb\backend\libusb1.py", line 954, in __read
    _check(retval)
  File "D:\TMP\project_venv_3.9\lib\site-packages\usb\backend\libusb1.py", line 604, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
usb.core.USBError: [Errno 32] Pipe error
```

What could I do to mitigate this problem?

Thank you!

**Comment by jonasmalacofilho** [comment, marked as answer]
Hi @danielhrisca,

Sorry for the delay, I almost lost the notification for this issue...

Please take a look at https://github.com/pyusb/pyusb/discussions/411#discussioncomment-2138235 for some general information about how to investigate a problem like this.  That comment deals with a different libusb API, so make sure to check the correct docs for the one being called here (`libusb_bulk_transfer`).

Also, issues like these usually arise from the device, and commonly require from you a better understanding of the protocol and requirements in use by the device.

---

P.S. Moving this to Discussions/Q/A.

**Reply by danielhrisca** [reply]
Could it be some kind of buffering problem (my application not reading fast enough from the USB pipe)?

**Reply by jonasmalacofilho** [reply]
I don't _think_ so.  I think it's more likely that you need to `write()` something (specific) before you can try to read from the device.

**Reply by danielhrisca** [reply]
I was probably not clear enough in the first post: 

* we have data acquisition device which has a custom command protocol over USB. The commands are send over one endpoint and the replies are received on the same endpoint. The acquired data is send by the device over a different endpoint. 
* there is no problem with the initialization phase: setting the configuration, setting up the acquisition details using the required command sequence
* the device starts to send the acquired data over the dedicated endpoint. For some period everything is OK, we continue to receive and process the incoming data stream. 
* after some time the read command on the dedicated stream fails with the IO error

This is why I was thinking about the possibility of a buffering problem. If we take ethernet for example, it is possible to configure the receive buffer size so that processing delay can be better tolerated. I could not find a similar setting for USB.

**Reply by jonasmalacofilho** [reply]
In that case, you should check the protocol documentation again, or get in touch with the manufacturer.

Just because the device reports the acquired data for some time, doesn't necessarily mean it will continue to do it forever: it might require a custom keepalive write, or something else along that line.  Again, it should be described in the protocol documentation, assuming it's not too incomplete/inaccurate.

In any case, this is not the timeout problem you describe.  Or, at the very least, the problem you describe is specifically documented by LibUSB to result in a different (timeout) error.[^1]

[^1]: https://libusb.sourceforge.io/api-1.0/group__libusb__syncio.html#ga2f90957ccc1285475ae96ad2ceb1f58c

**Comment by ARM-ASAP** [comment]
I have the same error, from console it works fine, from script it doesnt.
